### PR TITLE
[acr] Fix #14811 Add support for dockerignore override

### LINKED ISSUE #14811: dockerfile: add dockerignore override support
**Is your feature request related to a problem? Please describe.**

There is currently no way to specify (or override) the `.dockerignore` file in the `acr build` command. This prevents multiple Dockerfile(s) to share the same context effectively.

**Describe the solution you'd like**

I suggest that different Dockerfiles define different subsets of the context files.
The frontend (e.g. `acr build`) first checks for <path/to/Dockerfile>.dockerignore and if it is found it will be used instead of the root .dockerignore. 

This would be consistent with the docker behaviour implemented in:
https://github.com/moby/buildkit/pull/901

**Describe alternatives you've considered**

The alternative is to overwrite the .dockerignore file(s) at build time, only this prevents concurrent builds as there is no control over which .dockerignore file is used by which build thread.  Not that I can see at least

**Additional context**

It seems to me it would be fairly simple to modify:

https://github.com/Azure/azure-cli/blob/438cc0a9fb7f4ca2c99cadbfbc96cd7ea27fadaa/src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py#L147

to introduce this change. 

Would you accept a PR?

### PR DESCRIPTION
**Pull request by luigibk**
**Description**  

There is currently no way to specify (or override) the .dockerignore file in the acr build command. This prevents multiple Dockerfile(s) to share the same context effectively.

This change allows different Dockerfiles to define different subsets of the context files with different ignore rules.
In case the docker file is not named Dockerfile, the frontend (i.e. acr build) first checks for <path/to/docker_file_name>.dockerignore and if it is found it will be used instead of the root .dockerignore.

This is consistent with the docker behaviour implemented in:
moby/buildkit#901

For more information see the related issue #14811 

**Testing Guide**  
No command changes. This PR affects the `acr build` command, when the `-f <docker_file_name>` option is used. 

**History Notes**  

[acr] Add support for dockerignore override

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
acr

**Comment by yungezz** [comment]
hi @fengzhou-msft could you pls help to review this? thanks

**Comment by northtyphoon** [comment]
@shahzzzam @juliusl @rosanch

File: src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py
Comment (shahzzzam) [inline src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py:74]: where is `ignore_list_size` used? do you think its redundant to return that?

File: src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py
Comment (luigibk) [reply src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py:170]: Hi @juliusl, thanks for your comments. I had to exclude the `.dockerignore` file in case it's overridden as it gets picked up by the registry build task. Without doing this instead of overrding `.dockerignore` I would be merging the 2 dockerignore files.

File: src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py
Comment (luigibk) [reply src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py:73]: Yes, I can try and I agree with you that would be cleaner and more portable. I used this specific code (and comment) to be consistent with this: 

https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acr/build.py#L66

as I wanted to add this feature with the minimal possible change but the original code should be probably reviewed/refactored too.

File: src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py
Comment (luigibk) [reply src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py:74]: Hi @shahzzzam, thanks for your comment. It looks like it's used in:

https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acr/_archive_utils.py#L107

I am not sure if it would be appropriate to refactor that, but I personally see that as out of scope for this PR. Please let me know if you have suggestions in that sense.

**Comment by luigibk** [comment]
Hi @juliusl and @shahzzzam did you have a chance to have a look at my commit and replies to your comments?

**Comment by juliusl** [comment]
LGTM, LGTY @shahzzzam ?

**Review by juliusl — APPROVED** [review]

**Review by shahzzzam — APPROVED** [review]
LGTM

**Comment by luigibk** [comment]
Hi @northtyphoon, what's the next step for this? Am I waiting for another review?

**Comment by yungezz** [comment]
hi @fengzhou-msft, could you pls help to review the PR? thanks

**Comment by yungezz** [comment]
hi @fengzhou-msft, could you pls help to review the PR? thanks

**Review by fengzhou-msft — APPROVED** [review]

**Pull request merged**
